src/main/java/bio/terra/service/filedata/azure/util/BlobContainerClientFactory.java

```diff
@@ -0,0 +1,187 @@
+        createBlobServiceClientUsingSharedKey(accountName, accountKey)
+            .getBlobContainerClient(containerName);
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.SharedKey;
+    blobContainerSasTokenCreds = null;
+  }
+
+  public BlobContainerClientFactory(
+      String accountName, TokenCredential azureCredential, String containerName) {
+
+    if (StringUtils.isBlank(accountName)) {
+      throw new IllegalArgumentException("The account name is either null or empty.");
+    }
+
+    if (StringUtils.isBlank(containerName)) {
+      throw new IllegalArgumentException("The container name is either null or empty.");
+    }
+
+    blobServiceClient =
+        createBlobServiceClientUsingTokenCredentials(
+            accountName,
+            Objects.requireNonNull(azureCredential, "Azure token credentials are null."));
+    blobContainerClient = blobServiceClient.getBlobContainerClient(containerName);
+
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.UserDelegatedKey;
+    blobContainerSasTokenCreds = null;
+  }
+
+  public BlobContainerClientFactory(String containerURLWithSasToken) {
+
+    BlobUrlParts blobUrl = BlobUrlParts.parse(containerURLWithSasToken);
+
+    blobContainerSasTokenCreds =
+        new AzureSasCredential(blobUrl.getCommonSasQueryParameters().encode());
+    blobContainerClient =
+        new BlobContainerClientBuilder()
+            .httpClient(httpClient)
+            .endpoint(
+                String.format(
+                    Locale.ROOT,
+                    "https://%s/%s",
+                    blobUrl.getHost(),
+                    blobUrl.getBlobContainerName()))
+            .credential(blobContainerSasTokenCreds)
+            .buildClient();
+
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.ContainerSasToken;
+  }
+
+  public BlobContainerClient getBlobContainerClient() {
+    return blobContainerClient;
+  }
+
+  public String createReadOnlySasUrlForBlob(String blobName) {
+    if (sasTokenGeneratorStrategy.equals(SasTokenGeneratorStrategy.SharedKey)) {
+      return createReadOnlySasUrlForBlobUsingClient(blobName);
+    }
+
+    if (sasTokenGeneratorStrategy.equals(SasTokenGeneratorStrategy.ContainerSasToken)) {
+      return createReadOnlySasUrlForBlobUsingContainerSas(blobName);
+    }
+
+    if (sasTokenGeneratorStrategy.equals(SasTokenGeneratorStrategy.UserDelegatedKey)) {
+      return createReadOnlySasUrlForBlobUsingUserDelegatedSas(blobName);
+    }
+
+    throw new RuntimeException("Failed to generate SAS token for blob. Invalid strategy set.");
+  }
+
+  private synchronized String createReadOnlySasUrlForBlobUsingUserDelegatedSas(String blobName) {
+
+    if (delegationKey == null
+        || delegationKey.getSignedExpiry().isBefore(OffsetDateTime.now(ZoneOffset.UTC))) {
+      OffsetDateTime keyExpiry = OffsetDateTime.now(ZoneOffset.UTC).plusDays(EXPIRATION_DAYS);
+      delegationKey = blobServiceClient.getUserDelegationKey(null, keyExpiry);
```

Comment:
Is there a reason why `delegationKey` is cached after it's created? Is it because it's expensive to get the key every time? The reason for this cached data should be documented as it's not the typical way we use state in our objects, either here or next to the field definition. Putting this logic in a separate method would also work, as it's not dependent on `blobName` and could be written as `getDelegationKey()`.

Reply:
See above.